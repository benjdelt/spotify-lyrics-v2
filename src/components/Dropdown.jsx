import React, { useState, useEffect, useRef, Children } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretDown, faCaretUp } from '@fortawesome/free-solid-svg-icons'
import '../styles/Dropdown.css';


function DropdownMenuItem(props) {

  const handleClick = () => {
    props.handleClick();
    props.openHandler();
  }

  return (
    <div key={props.key} className="dropdown-option" onClick={handleClick} >
      <span className="dropdown-option-text">{props.children}</span>
    </div>
  );
}

function Dropdown(props) {

  const wrapperRef = useRef(null);

  const [open, setOpen] = useState(false);

  useEffect(() => {
    document.addEventListener("click", handleClickOutside, false);
    return () => {
      document.removeEventListener("click", handleClickOutside, false);
    };
  }, [])

  const handleClick = event => {
    event.stopPropagation();
    setOpen(!open);
  };

  const handleClickOutside = event => {
    if (wrapperRef.current && !wrapperRef.current.contains(event.target)) {
      setOpen(false);
    }
  };

  const openHandler = () => {
    setOpen(false);
  };

  const closedIcon = props.up ? faCaretUp : faCaretDown;
  const openIcon = props.up ? faCaretDown : faCaretUp;

  const styles = props.up ? {
    visibility: open ? "visible" : "hidden",
    position: "relative",
    bottom: "6.5em",
  } : {
    visibility: open ? "visible" : "hidden",
  };

  return (
    <div className="dropdown-parent">
      <button onClick={handleClick} className="dropdown-label">
        {props.label} &nbsp;
        <FontAwesomeIcon icon={open ? openIcon : closedIcon} />
      </button>
      <div className="dropdown-menu" style={styles} ref={wrapperRef}>
        {Children.map(props.children, child => React.cloneElement(child, { openHandler }))}
      </div>
    </div>
  )
}

export { Dropdown, DropdownMenuItem };
