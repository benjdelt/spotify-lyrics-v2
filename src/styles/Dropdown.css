.dropdown-parent:hover, .dropdown-parent:focus {
  text-decoration: none;
}

.dropdown-label:hover, .dropdown-label:focus {
  text-decoration: underline;
}

.dropdown-label {
  display: flex;
  align-items: center;
}

.dropdown-menu {
  background-color: #1c1c1c;
  margin: .2em;
  padding: .5em; 
}

.dropdown-option {
  padding: .2em;
  cursor: pointer;
}

.dropdown-option:hover, .dropdown-option:focus {
  background-color: #3e3e3e;
  color: #fff;
}