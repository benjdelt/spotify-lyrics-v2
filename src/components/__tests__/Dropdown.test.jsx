import { render, cleanup } from '@testing-library/react';
import { Dropdown } from '../Dropdown';


afterEach(cleanup)

it('should take a snapshot', () => {
  const { asFragment } = render(<Dropdown />) 
  expect(asFragment(<Dropdown />)).toMatchSnapshot()
});
