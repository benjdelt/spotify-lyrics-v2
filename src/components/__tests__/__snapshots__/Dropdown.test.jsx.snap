// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown-parent"
  >
    <button
      class="dropdown-label"
    >
       Â 
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 "
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </button>
    <div
      class="dropdown-menu"
      style="visibility: hidden;"
    />
  </div>
</DocumentFragment>
`;
